detectron2/export/patch_instance.py

```diff
@@ -0,0 +1,88 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
+
+import importlib.util
+import os
+import sys
+import tempfile
+import torch
+
+from detectron2.structures import Boxes, Instances  # noqa F401
+
+_counter = 0  # TODO replace by a hash?
+
+
+def export_torchscript_with_patch_instance(model, fields):
+    with tempfile.NamedTemporaryFile(mode="w", encoding="utf-8", suffix=".py", delete=False) as f:
+        cls_name, s = gen_class(fields)
+        f.write(s)
+        f.flush()
+        f.close()
+
+        module = _import(f.name)
+        new_instance = getattr(module, cls_name)
+        _ = torch.jit.script(new_instance)
+
+        Instances.__torch_script_class__ = True
+        Instances._jit_override_qualname = torch._jit_internal._qualified_name(new_instance)
+        return torch.jit.script(model), new_instance
```

Comment:
what's the use case to return `new_instance`?

Reply:
Used in the UT of RPN. Before executing the forward function of scripted_RPN class, it seems that JIT will 'inspect' all the attributes of scripted_Instance (e.g. the new Instance) like `scripted_Instance.proposal_boxes`. If we pass origin Instance to scripted_RPN, `__getattr__` of origin Instance class will be executed. But in the meantime, `proposal_boxes` is not exist in the `_fields`, AttributeError will be raised.

Reply:
I see. I think there are two issues to addressed:
1. We should only test for inference mode in the unittest for now. Most of the training codepath is unsupported anyway.
2. Even in inference mode, a unittest might still need to pass `Instances` as input (e.g. in `ROIHeads`). However I think this use case is almost only used by unittest but not by users. So we should provide to users a simpler API.

With the above suggestion on contextmanager, we can use 
```
with patch_instance(...) as new_instance:
    s = torch.jit.script(model)
```
in unittests, but package them together to users in a function without the extra return value that they don't use.

Reply:
done.  `export_torchscript_with_patch_instance` only returns torchscript now.